### ISSUE TITLE
[1.8.9] Optifine HD U M5 resets FPS cap at every launch of the game

### ISSUE THREAD

**Issue by EktoHunter** [sp614x/optifine#5625]
## Description of Issue
Whenever playing Minecraft using this version of Optifine is Launched, the game will set the maxFPS value in Options.txt to 260.
Multiple people report this issue, it never seems to have been fixed:
[1](https://github.com/sp614x/optifine/issues/4356) [2](https://www.reddit.com/r/Optifine/comments/garf1u/my_fps_cap_gets_reset_but_nothing_else/) [3](https://github.com/sp614x/optifine/issues/5168) [4](https://www.minecraftforum.net/forums/support/java-edition-support/1953277-fpslimit-resetting)

## Steps to Reproduce
1. Set the FPS cap to anything except VSYNCH, for example 90 in-game.
2. Close the game (Doesn't matter whether its using the "X" for the windows menu or the "quit game" button)
3. Open up the Minecraft launcher and launch the game again
4. The FPS will have reset to "Unlimited".

Alternatively, if you change the same setting (maxFPS:XX) in "%appdata%\.minecraft\options.txt" the same will occur.

## OptiFine Version
Optifine HD U M5 1.8.9

## Installation Method
Standalone Installer

## Other Installed Mods
None

## Log Files/Crash Reports
[latest.log](https://github.com/sp614x/optifine/files/6443831/latest.log)

## F3 Debug Screenshot
![image](https://user-images.githubusercontent.com/23218251/117498210-ca9b2180-af79-11eb-9535-29d658e013bf.png)

## Prior Testing
This only happens with Optifine in OptiFine standalone, doesn't happen in Vanilla or 1.16.5 Optifine

## Additional Information
_When the FPS setting is unlimited I sometimes worry that my PC is gonna take off with how loud the fan gets._ 

Edit: Some more Info on my PC/Laptop if that's helpful:
![image](https://user-images.githubusercontent.com/23218251/117499270-46e23480-af7b-11eb-82ef-0fb459868ea0.png)

![image](https://user-images.githubusercontent.com/23218251/117499245-3df16300-af7b-11eb-9334-1e57b34b6b68.png)

**Comment by anzz1** [comment]
Still happens on both "U M5" and "U M6 pre2", in a fresh vanilla installation.

**Comment by anzz1** [comment]
I figured out the reason, it's a logic error in "avh.class" where the options are parsed. This part particularly:
```java
if (astring[0].equals("maxFps")) {
    this.g = Integer.parseInt(astring[1]);
    if (this.t) {
        this.g = (int)avh.a.i.f();
    }
    if (this.g <= 0) {
        this.g = (int)avh.a.i.f();
    }
}
```

`this.t` is the enableVsync option, which is set to true by default.
The problem is that the options are parsed in a per-line basis from the options.txt , and ´maxFps´ option is written *before* ´enableVsync´ option. Thus when the logic reaches maxFps option, enableVsync will always be the default value , true, since it hasn't reached the option yet.

You can easily test this by manually editing options.txt, moving the `enableVsync:false` option to before `maxFps:200`. It will now load the correct configuration, but after exiting the config is saved in the wrong order again.

**Comment by anzz1** [comment]
A hacky way to fix it is by opening "OptiFine-1.8.9_HD_U_M6_pre2.jar" and extracting "avh.class" out. Then open the class with a HEX editor, change byte 04 -> 03 at these two locations: 0x5EFB and 0x68CF. Put the avh.class back in to the jar.
This will change that enableVsync = false by default, so the config loads properly now.

**Comment by Muffus16** [comment]
> 

0x5EFB shows the byte 0x21 for me.
Is it possible that the position isnt quite correct?

**Comment by anzz1** [comment]
> > 
> 
> 0x5EFB shows the byte 0x21 for me. Is it possible that the position isnt quite correct?

![image](https://github.com/sp614x/optifine/assets/13628128/336935f0-4ea0-4395-b962-99fd2ca9f731)

**Comment by anzz1** [comment]
> Is it possible, that the file has changed?
> Could you reinstall of 1.8.9 m6 pre2 and check again?
> Thanks.
Nope, you have bunged something up. Fear not though, you can also change the corresponding .xdelta file in the installer.

1. Download [preview_OptiFine_1.8.9_HD_U_M6_pre2.jar](https://optifine.net/adloadx?f=preview_OptiFine_1.8.9_HD_U_M6_pre2.jar&x=91b8)
2. Extract `avh.class.xdelta` from the "patch" folder inside the .jar file.
3. Change bytes:
  - 0x4D44 : 04 -> 03
  - 0x5722 : 04 -> 03
4. Put `avh.class.xdelta` back to the jar, replacing the original one.
5. Proceed installation of OptiFine as usual.

Didn't test it, but should work. Godspeed.

**Comment by Muffus16** [comment]
> > Is it possible, that the file has changed?
> > Could you reinstall of 1.8.9 m6 pre2 and check again?
> > Thanks.
> 
> Nope, you have bunged something up. Fear not though, you can also change the corresponding .xdelta file in the installer.
> 
> 1. Download [preview_OptiFine_1.8.9_HD_U_M6_pre2.jar](https://optifine.net/adloadx?f=preview_OptiFine_1.8.9_HD_U_M6_pre2.jar&x=91b8)
> 2. Extract `avh.class.xdelta` from the "patch" folder inside the .jar file.
> 3. Change bytes:
> 
> * 0x4D44 : 04 -> 03
> * 0x5722 : 04 -> 03
> 
> 4. Put `avh.class.xdelta` back to the jar, replacing the original one.
> 5. Proceed installation of OptiFine as usual.
> 
> Didn't test it, but should work. Godspeed.

Now comes the interesting part.
I did change it like you said in the instller jar file.
yes these two 0x04 byte were existing.

Installer returned not matching md5 sum wich totally makes sense caclulted sum saved it to avh.class.md5
same error. but it also displays calculated sum which differs from my caclulated sum.
I think it is because it merged code with original game code.

**Long story short it did work thanks alot.**

But if im now checking on the file avh.class it does actually show me 0x21 for byte 0x5efb
which means there must be a difference in our avh.class files.
if the installer merges the code it could be that your vanilla 1.8.9 files are not matching mine.
at least thats what i think could have happend here.

**Comment by anzz1** [comment]
Glad that you got it sorted out.

**Comment by EktoFox** [comment]
I don't think that this version of optifine will still be updated, so this will probably stay a permanent issue for anyone downloading this build of optifine.

**Comment by Muffus16** [comment]
Do you still need a patch for your game?
